### PR TITLE
#4556 fix: `self-signature` key import issue in Add key screen 

### LINKED ISSUE #4556: Error "Could not find valid self-signature in key in key-id"
Description:
An enterprise customer receives the following error: Could not find valid self-signature in \<key-id> when trying to use a certain private key. I check upon the key packets and it doesn't have a "Signature Packet" along with the other packets. That's all I've noticed so far so I tried reading the private key through custom code and got the same error. 

image

<details><summary>dumped key info (redacted):</summary>

```
Old: Secret Key Packet(tag 5)(1862 bytes)
	Ver 4 - new
	Public key creation time - Tue May  2 12:20:12 PST 2017
	Pub alg - RSA Encrypt or Sign(pub 1)
	RSA n(4096 bits) - ...
	RSA e(17 bits) - ...
	Sym alg - AES with 256-bit key(sym 9)
	Iterated and salted string-to-key(s2k 3):
		Hash alg - SHA1(hash 2)
		Salt - b7 34 33 89 6e b1 6e b8 
		Count - 4063232(coded count 191)
	IV - 8a 92 5c 56 80 e6 1d c7 c9 7f 00 b6 31 32 42 4e 
	Encrypted RSA d
	Encrypted RSA p
	Encrypted RSA q
	Encrypted RSA u
	Encrypted SHA1 hash
Old: User ID Packet(tag 13)(51 bytes)
	User ID - <userid-one> (JV) <redacted@email.com>
Old: User ID Packet(tag 13)(39 bytes)
	User ID - <userid-two> <redacted@email.com>
Old: User ID Packet(tag 13)(41 bytes)
	User ID - <userid-three> <redacted@email.com>
Old: Secret Subkey Packet(tag 7)(1862 bytes)
	Ver 4 - new
	Public key creation time - Tue May  2 12:20:14 PST 2017
	Pub alg - RSA Encrypt or Sign(pub 1)
	RSA n(4096 bits) - ...
	RSA e(17 bits) - ...
	Sym alg - AES with 256-bit key(sym 9)
	Iterated and salted string-to-key(s2k 3):
		Hash alg - SHA1(hash 2)
		Salt - f7 1a 1f ac c1 52 cf db 
		Count - 4063232(coded count 191)
	IV - f6 ae 89 43 05 a0 da 1e b7 9b ff f6 f5 38 04 57 
	Encrypted RSA d
	Encrypted RSA p
	Encrypted RSA q
	Encrypted RSA u
	Encrypted SHA1 hash

```
</details>

Reference: https://mail.google.com/mail/u/human@flowcrypt.com/#inbox/FMfcgzGpGwtXHZPgFNwtFHzJnXBQhChV

### PR DESCRIPTION
**Pull request by ioanmo226**
This PR fixed `self-signature` key import issue in Add key screen 

close #4556 // if this PR closes an issue

----------------------------------

**Tests** _(delete all except exactly one)_:
- Tests added or updated

--------------------------------

### To be filled by reviewers

I have reviewed that this PR... _(tick whichever items you personally focused on during this review)_:
- [x] addresses the issue it closes (if any)
- [x] code is readable and understandable
- [x] is accompanied with tests
- [x] is free of vulnerabilities
- [x] doesn't need documentation

### CODE REVIEW

File: test/source/tests/setup.ts [hunk 1]
```diff
@@ -497,6 +437,28 @@ AN8G3r5Htj8olot+jm9mIa5XLXWzMNUZgg==
...
+      await addKeyPopup.waitAll('@input-compatibility-fix-expire-years', { timeout: 30 });
+      await addKeyPopup.selectOption('@input-compatibility-fix-expire-years', '1');
+      await addKeyPopup.waitAndClick('@action-fix-and-import-key');
+      await Util.sleep(1);
+      const myKeyFrame = await SettingsPageRecipe.awaitNewPageFrame(settingsPage, `@action-show-key-2`, ['my_key.htm', 'placement=settings']);
+      await Util.sleep(1);
+      const curDate = new Date();
+      // getFullYear() + 1 because we set expiration as 1 year
+      const expiration = `${curDate.getFullYear() + 1}-${curDate.getMonth() + 1}-${curDate.getDate()}`;
+      expect(await myKeyFrame.read('@content-key-expiration')).to.equal(expiration);
```
Comment (rrrooommmaaa) [inline test/source/tests/setup.ts:459]: This test (or the key-fixing code) is susceptible to the time zone setting and/or date boundary.
image
image

=> AssertionError: expected '2023-11-16' to equal '2023-11-15'

Comment (rrrooommmaaa) [reply test/source/tests/setup.ts:459]: image

Comment (rrrooommmaaa) [reply test/source/tests/setup.ts:459]: image

Comment (ioanmo226) [reply test/source/tests/setup.ts:459]: I honestly don't like current way of formatting date.
https://github.com/FlowCrypt/flowcrypt-browser/blob/128b3ea4c5207827225bd37e92ea1853cb3ea1aa/extension/chrome/settings/modules/my_key.ts#L58
image

Comment (ioanmo226) [reply test/source/tests/setup.ts:459]: We may use `date-fns` (or other library like that) or make common `format` function which accepts `format` string.

Comment (ioanmo226) [reply test/source/tests/setup.ts:458]: For now let me change above test code to below line (to let test pass despite of timezones)
But I suggest to implement date format functions like above comment and make new issue/PR for that.

```ts
const expirationDate = new Date(new Date().setFullYear(new Date().getFullYear() + 1));
const expiration = Str.datetimeToDate(Str.fromDate(expirationDate));
```

Comment (ioanmo226) [reply test/source/tests/setup.ts:458]: What do you think @rrrooommmaaa ?

Comment (rrrooommmaaa) [reply test/source/tests/setup.ts:458]: > For now let me change above test code to below line (to let test pass despite of timezones)

It's certainly better, but may fail occasionally if the test is run at about 23:59 UTC
AssertionError: expected '2023-11-16' to equal '2023-11-17' (we can leave a leeway of minus one day) 

> But I suggest to implement date format functions like above comment and make new issue/PR for that.

For `dependencies` I'd ask for @tomholub 's approval.
For `devDependencies` we have more freedom )

Comment (ioanmo226) [reply test/source/tests/setup.ts:458]: ok

File: test/source/tests/setup.ts [hunk 2]
```diff
@@ -497,6 +437,32 @@ AN8G3r5Htj8olot+jm9mIa5XLXWzMNUZgg==
...
+      await Util.sleep(1);
+      const curDate = new Date(), year = curDate.getFullYear(), month = curDate.getMonth(), date = curDate.getDate();
+      const expirationDate = new Date(year + 1, month, date);
+      // Had to add this because if test runs at 23:59:59 it might cause assertion error
+      // https://github.com/FlowCrypt/flowcrypt-browser/pull/4796#discussion_r1025150001
+      const oneDayAfterExpirationDate = new Date(year + 1, month, date + 1);
+      const expiration = Str.datetimeToDate(Str.fromDate(expirationDate));
+      const oneDayAfterExpiration = Str.datetimeToDate(Str.fromDate(oneDayAfterExpirationDate));
+      expect(await myKeyFrame.read('@content-key-expiration')).to.be.oneOf([expiration, oneDayAfterExpiration]);
+    }));
```
Comment (rrrooommmaaa) [inline test/source/tests/setup.ts:464]: Now you're comparing `2023-11-16' (created at 23:59) against ['2023-11-17', '2023-11-18'].
Should be [oneDayBeforeExpectedExpiration, expectedExpiration]

File: test/source/tests/setup.ts [hunk 3]
```diff
@@ -497,6 +437,32 @@ AN8G3r5Htj8olot+jm9mIa5XLXWzMNUZgg==
...
+      await addKeyPopup.selectOption('@input-compatibility-fix-expire-years', '1');
+      await addKeyPopup.waitAndClick('@action-fix-and-import-key');
+      await Util.sleep(1);
+      const myKeyFrame = await SettingsPageRecipe.awaitNewPageFrame(settingsPage, `@action-show-key-2`, ['my_key.htm', 'placement=settings']);
+      await Util.sleep(1);
+      const curDate = new Date(), year = curDate.getFullYear(), month = curDate.getMonth(), date = curDate.getDate();
+      const expirationDate = new Date(year + 1, month, date);
+      // Had to add this because if test runs at 23:59:59 it might cause assertion error
+      // https://github.com/FlowCrypt/flowcrypt-browser/pull/4796#discussion_r1025150001
+      const oneDayAfterExpirationDate = new Date(year + 1, month, date + 1);
```
Comment (rrrooommmaaa) [inline test/source/tests/setup.ts:460]: will it handle `date` overflow/underflow correctly?
I think it's ok to use the package you suggested for the test project (if it helps to simplify this)

**Review by rrrooommmaaa — APPROVED** [review]

**Pull request merged**